topydo/ui/columns/Main.py

```diff
@@ -269,38 +288,44 @@ def _execute_handler(self, p_command, p_todo_id=None, p_output=None):
 
         self._last_cmd = (p_command, p_output == self._output)
 
-        if '{}' in p_command:
-            if self._has_marked_todos():
-                p_todo_id = ' '.join(self.marked_todos)
-            p_command = p_command.format(p_todo_id)
-
         try:
             p_command = shlex.split(p_command)
         except ValueError as verr:
             self._print_to_console('Error: ' + str(verr))
             return
 
         try:
-            (subcommand, args) = get_subcommand(p_command)
+            subcommand, args = get_subcommand(p_command)
         except ConfigError as cerr:
             self._print_to_console(
                 'Error: {}. Check your aliases configuration.'.format(cerr))
             return
 
-        self._backup(subcommand, args)
+        env_args = (self.todolist, p_output, self._output, self._input)
 
-        try:
-            command = subcommand(
-                args,
-                self.todolist,
-                p_output,
-                self._output,
-                self._input,
-            )
+        if '{}' in args:
+            if self._has_marked_todos():
+                ids = self.marked_todos
+            else:
+                ids = {p_todo_id} if p_todo_id else set()
 
-            if command.execute() != False:
-                self._post_execute()
+            invalid_ids = self._check_id_validity(ids)
+
+            if invalid_ids:
+                self._print_to_console('Error: ' + invalid_ids)
+                return
 
+            transaction = Transaction(subcommand, env_args, ids)
+        else:
+            transaction = Transaction(subcommand, env_args)
+
+        transaction.prepare(args)
+        label = transaction.label
+        self._backup(subcommand, p_label=label)
+
+        try:
+            if transaction.execute():
```

Comment:
Inside a transaction, where the first command succeeds and the second command fails (its `execute()` method returns `False` for whatever reason), it would leave an inconsistent state. The `TodoList` in memory was modified but `self._post_execute()` would not be called (and therefore the first change is not flushed to disk).

I think a transaction needs to store a copy of the `TodoList` instance before it starts executing, and restore that copy whenever a command fails (so the net effect is that Transaction didn't do anything).